plugin/trino-hive/src/main/java/io/trino/plugin/hive/HiveMetadata.java

```diff
@@ -1345,6 +1354,79 @@ public void dropColumn(ConnectorSession session, ConnectorTableHandle tableHandl
+    {
+        HiveTableHandle table = (HiveTableHandle) tableHandle;
+        failIfAvroSchemaIsSet(table);
+        HiveColumnHandle column = (HiveColumnHandle) columnHandle;
+
+        table.getPartitionNames().ifPresent(partitionNames -> {
+            if (partitionNames.contains(column.getName())) {
+                throw new TrinoException(NOT_SUPPORTED, "Changing partition column types is not supported");
+            }
+        });
+        if (!canChangeColumnType(column.getType(), type)) {
+            throw new TrinoException(NOT_SUPPORTED, "Cannot change type from %s to %s".formatted(column.getType(), type));
+        }
+        metastore.setColumnType(table.getSchemaName(), table.getTableName(), column.getName(), toHiveType(type));
+    }
+
+    private static boolean canChangeColumnType(Type sourceType, Type targetType)
+    {
+        if (sourceType.equals(targetType)) {
+            return true;
+        }
+        if (sourceType == TINYINT) {
+            return targetType == SMALLINT || targetType == INTEGER || targetType == BIGINT;
+        }
+        if (sourceType == SMALLINT) {
+            return targetType == INTEGER || targetType == BIGINT;
+        }
+        if (sourceType == INTEGER) {
+            return targetType == BIGINT;
+        }
+        if (sourceType == REAL) {
+            return targetType == DOUBLE;
+        }
+        if (sourceType instanceof VarcharType || sourceType instanceof CharType) {
+            return targetType instanceof VarcharType || targetType instanceof CharType;
+        }
+        if (sourceType instanceof DecimalType sourceDecimal && targetType instanceof DecimalType targetDecimal) {
+            // TODO: Support rescale in ORC DecimalColumnReader
+            return sourceDecimal.getScale() == targetDecimal.getScale()
+                    && sourceDecimal.getPrecision() <= targetDecimal.getPrecision();
+        }
+        if (sourceType instanceof ArrayType sourceArrayType && targetType instanceof ArrayType targetArrayType) {
+            return canChangeColumnType(sourceArrayType.getElementType(), targetArrayType.getElementType());
+        }
+        if (sourceType instanceof RowType sourceRowType && targetType instanceof RowType targetRowType) {
+            List<RowType.Field> fields = Streams.concat(sourceRowType.getFields().stream(), targetRowType.getFields().stream())
+                    .distinct()
+                    .collect(toImmutableList());
+            for (RowType.Field field : fields) {
+                String fieldName = field.getName().orElseThrow();
+                boolean allowedChange = findFieldByName(sourceRowType.getFields(), fieldName)
+                        .flatMap(sourceField -> findFieldByName(targetRowType.getFields(), fieldName)
+                                .map(targetField -> canChangeColumnType(sourceField.getType(), targetField.getType())))
+                        // Allow removal and addition of fields
```

Comment:
By removing and adding a field back (same name, different type), I can effectively change type of any field, right?
(is it same for columns? i could drop and re-add the column)

Expand the comment with the reasoning.